### ISSUE TITLE
tesseract-recognize creates negative word coordinates

### ISSUE THREAD

**Issue by stweil** [OCR-D/ocrd_tesserocr#153]
In a [workflow with PPN1024726142](https://github.com/OCR-D/ocrd_tesserocr/issues/149#issuecomment-691657163), `tesseract-recognize` created a negative coordinate for page 11:

             <pc:TextLine id="region0008_line0000">
                 <pc:Coords points="566,1558 0,1539 0,1565 564,1583"/>
                 <pc:Word id="region0008_line0000_word0000">
                    <pc:Coords points="0,1546 566,1565 565,1584 -1,1565"/>

`ocrd-transform` fails to process that page without an error message when converting from PAGE to ALTO.

**Comment by bertsky** [comment]
> `tesseract-recognize` created a negative coordinate for page 11:

Yes, we should use the `polygon_for_parent` mechanism for word and glyph segmentation, too.

> `ocrd-transform` fails to process that page without an error message when converting from PAGE to ALTO.

Cannot speak to that.

**Comment by stweil** [comment]
The negative x coordinate causes am exception in `PageConverter`:

    $ ocr-transform page alto bad.xml /tmp/bad.xml
    Exception in thread "main" java.lang.NullPointerException
            at org.primaresearch.dla.page.converter.PageConverter.handleNegativeCoordinates(PageConverter.java:389)
            at org.primaresearch.dla.page.converter.PageConverter.run(PageConverter.java:216)
            at org.primaresearch.dla.page.converter.PageConverter.main(PageConverter.java:130)

**Comment by bertsky** [comment]
> > `tesseract-recognize` created a negative coordinate for page 11:
> 
> Yes, we should use the `polygon_for_parent` mechanism for word and glyph segmentation, too.

Should be covered by #152 now, too (sorry for inconvenience).

**Comment by stweil** [comment]
It now no longer creates negative word coordinates. I still get the `java.lang.NullPointerException` from `PageConverter`, but now for page 43 (which has no text and no word coordinates), so that seems to be unrelated.

**Comment by stweil** [comment]
@bertsky, @kba, the remaining problem with `PageConverter` occurs when converting pages without any text from PAGE to ALTO. Currently the PAGE XML of an empty page contains an entry `<pc:ReadingOrder/>`. This empty reading order obviously confuses `PageConverter`. I don't know whether an empty reading order is valid PAGE XML. Maybe it would be best to simply not write such an entry. See also https://github.com/PRImA-Research-Lab/prima-page-converter/issues/15.

**Comment by bertsky** [comment]
> the remaining problem with `PageConverter` occurs when converting pages without any text from PAGE to ALTO. Currently the PAGE XML of an empty page contains an entry `<pc:ReadingOrder/>`. This empty reading order obviously confuses `PageConverter`. I don't know whether an empty reading order is valid PAGE XML. Maybe it would be best to simply not write such an entry.

It's invalid by the schema, that's why the converter fails. We have already been trying to avoid these circumstances, but in case of `ocrd-tesserocr-segment-region` we forgot to remove empty `ReadingOrder` elements.

@kba do you think we could easily solve that for all processors at once with a fix during serialization in core? The alternative would be adding the following to all segmentation processors (right before `to_xml`):
```python
    ro = pcgts.get_Page().get_ReadingOrder()
    if ro and not ro.get_OrderedGroup() and not ro.get_UnorderedGroup():
        pcgts.get_Page().set_ReadingOrder(None)
```

**Comment by kba** [comment]
> @kba do you think we could easily solve that for all processors at once with a fix during serialization in core? 

Yes. https://github.com/OCR-D/core/issues/602

**Comment by bertsky** [comment]
> > @kba do you think we could easily solve that for all processors at once with a fix during serialization in core?
> 
> Yes. [OCR-D/core#602](https://github.com/OCR-D/core/issues/602)

Thanks!

Closing this issue – solved by #152
